plugins/com.google.cloud.tools.eclipse.appengine.deploy/src/com/google/cloud/tools/eclipse/appengine/deploy/AppEngineStandardDeployCommandHandler.java

```diff
@@ -3,17 +3,94 @@
+                                        FacetedProjectHelper facetedProjectHelper) {
+    if (exporter == null) {
+      projectToStagingExporter = new ProjectToStagingExporter();
+    } else {
+      projectToStagingExporter = exporter;
+    }
+    if (stagingDirectoryProvider == null) {
+      this.stagingDirectoryProvider = new DialogStagingDirectoryProvider();
+    } else {
+      this.stagingDirectoryProvider = stagingDirectoryProvider;
+    }
+    if (facetedProjectHelper == null) {
+      this.facetedProjectHelper = new FacetedProjectHelper();
+    } else {
+      this.facetedProjectHelper = facetedProjectHelper;
+    }
+  }
+  
   @Override
   public Object execute(ExecutionEvent event) throws ExecutionException {
-    new MessageDialog(HandlerUtil.getActiveShell(event), Messages.getString("deploy.standard.dialog.title"), null, //$NON-NLS-1$
-      Messages.getString("deploy.standard.dialog.message"), //$NON-NLS-1$
-      MessageDialog.INFORMATION, new String[]{Messages.getString("button.ok.label")}, 0).open(); //$NON-NLS-1$
-    return null;
+    try {
+      IEvaluationContext applicationContext = (IEvaluationContext)event.getApplicationContext();
+      //TODO should we use the approach from http://www.programcreek.com/2013/02/eclipse-rcp-tutorial-add-a-menu-item-to-the-popup-menu-when-right-click-a-file/
+      IStructuredSelection selection = 
+          (IStructuredSelection) applicationContext.getVariable(ISources.ACTIVE_CURRENT_SELECTION_NAME);
+      if (selection.getFirstElement() instanceof IProject) {
+        IProject project = (IProject) selection.getFirstElement();
+        IFacetedProject facetedProject = facetedProjectHelper.getFacetedProject(project);
+        // TODO replace with constant from AppEngineFacet (after possibly relocating that class)
+        if (facetedProjectHelper.projectHasFacet(facetedProject, 
+                                                 "com.google.cloud.tools.eclipse.appengine.facet")) {
+          exportProjectToStageDir(event, project);
+          //TODO run stage and deploy operations
+        }
+      }
+      // must be null, reserved for future use
+      return null;
+    } catch (CoreException coreException) {
+      throw new ExecutionException("Failed to export the project as exploded WAR", coreException);
+    }
+  }
+
+  private void exportProjectToStageDir(ExecutionEvent event, IProject iProject) throws CoreException {
+    String stageDir = getStagingDirectory(event);
+    if (stageDir != null) {
+      writeProjectToStageDir(iProject, stageDir);
+      verifyProjectInStageDir(stageDir);
+    }
+  }
+
+  private void writeProjectToStageDir(IProject project, String stageDir) throws CoreException {
+    projectToStagingExporter.writeProjectToStageDir(project, stageDir);
+  }
+
+  private String getStagingDirectory(ExecutionEvent event) {
+    return stagingDirectoryProvider.getStagingDirectory(event);
```

Comment:
`ExecutionEvent` objects aren't meant to be passed around; your objects independent of the Commands framework (`ExecutionEvent`)